support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala

```diff
@@ -112,61 +115,64 @@ object DigitalSubscriptionEmailAttributes {
-      case (Purchase(paymentInfo), ReaderType.Gift) =>
+  def build(digi: DigitalSubscriptionCreated)(implicit ec: ExecutionContext): Future[List[EmailFields]] =
+    digi match {
+      case giftPurchase: DigitalSubscriptionGiftPurchaseCreated =>
         for {
-          giftRecipient <- maybeGiftPurchase.toRight("Gift redemption must have a gift recipient")
           emails <- List(
-            giftPurchaserConfirmation(paymentInfo, giftRecipient),
-            giftRecipientNotification(giftRecipient)
+            giftPurchaserConfirmation(giftPurchase),
+            Future.successful(giftRecipientNotification(giftPurchase))
           ).sequence
         } yield emails
-      case (Purchase(paymentInfo), _) => directThankYou(paymentInfo).map(List(_))
-      case (Redemption, ReaderType.Corporate) => corpRedemption.map(List(_))
-      case (Redemption, ReaderType.Gift) => giftRedemption.map(List(_))
-      case (Redemption, _) => Left("redemption is only possible for gift and corporate subs")
+      case directPurchase: DigitalSubscriptionDirectPurchaseCreated => directThankYou(directPurchase).map(List(_))
+      case DigitalSubscriptionCorporateRedemptionCreated(_, subscriptionNumber) => Future.successful(List(corpRedemption(subscriptionNumber)))
+      case DigitalSubscriptionGiftRedemptionCreated(product) => Future.successful(List(giftRedemption(product.billingPeriod)))
     }
-  }
 
-  private def wrap(dataExtensionName: String, fields: DigitalSubscriptionEmailAttributes) = for {
-    attributePairs <- JsonToAttributes.asFlattenedPairs(fields.asJsonObject)
-  } yield EmailFields(attributePairs, Left(sfContactId), user.primaryEmailAddress, dataExtensionName)
+  private def wrap(dataExtensionName: String, fields: DigitalSubscriptionEmailAttributes): EmailFields = {
+    val attributePairs = JsonToAttributes.asFlattenedPairs(fields.asJsonObject).left.map(
+      error => throw new RuntimeException(s"coding error: $error")
+    ).merge
+    EmailFields(attributePairs, Left(sfContactId), user.primaryEmailAddress, dataExtensionName)
+  }
 
-  private def giftRecipientNotification(giftPurchase: DigitalSubscriptionGiftPurchase) =
+  private def giftRecipientNotification(giftPurchase: DigitalSubscriptionGiftPurchaseCreated) =
     wrap("digipack-gift-notification", GifteeNotificationAttributes(
       gifter_first_name = user.firstName,
       gift_personal_message = giftPurchase.giftRecipient.message,
       gift_code = giftPurchase.giftCode.value,
       last_redemption_date = formatDate(giftPurchase.lastRedemptionDate),
     ))
 
-  private def giftPurchaserConfirmation(paymentMethodWithSchedule: PaymentMethodWithSchedule, giftPurchase: DigitalSubscriptionGiftPurchase) = {
+  private def giftPurchaserConfirmation(giftPurchase: DigitalSubscriptionGiftPurchaseCreated)(implicit ec: ExecutionContext) = {
     import giftPurchase._
+    import purchaseInfo._
+
+    val promotion = paperFieldsGenerator.getAppliedPromotion(
+      promoCode,
```

Comment:
I'm not sure if gift purchases can have a promo code actually?

Reply:
They can't currently

Reply:
they are now added to the code just in case we need them in future.